diff --git a/dist/js/edit_form.js b/dist/js/edit_form.js
index 754b6a7..0ba6d8c 100644
--- a/dist/js/edit_form.js
+++ b/dist/js/edit_form.js
@@ -1,13 +1,21 @@
 (Drupal => {
   Drupal.behaviors.ginEditForm = {
-    attach: context => {
+    attach: function (context, settings) {
       once("ginEditForm", ".region-content form.gin-node-edit-form", context).forEach((form => {
         const sticky = context.querySelector(".gin-sticky"), newParent = context.querySelector(".region-sticky__items__inner");
         if (newParent && 0 === newParent.querySelectorAll(".gin-sticky").length) {
           newParent.appendChild(sticky);
           const actionButtons = newParent.querySelectorAll("button, input, select, textarea");
           actionButtons.length > 0 && actionButtons.forEach((el => {
-            el.setAttribute("form", form.getAttribute("id")), el.setAttribute("id", el.getAttribute("id") + "--gin-edit-form");
+            let element_id = el.getAttribute("id");
+            let new_id = element_id + "--gin-edit-form";
+            el.setAttribute("form", form.getAttribute("id")), el.setAttribute("id", new_id);
+
+            if (settings.ajax && settings.ajax.hasOwnProperty(element_id)) {
+              settings.ajax[new_id] = settings.ajax[element_id];
+              settings.ajax[new_id].selector = '#' + new_id;
+              delete settings.ajax[element_id];
+            }
           }));
         }
       }));
diff --git a/js/edit_form.js b/js/edit_form.js
index 4ff4e18..16ddfb2 100644
--- a/js/edit_form.js
+++ b/js/edit_form.js
@@ -4,7 +4,7 @@
 
 ((Drupal) => {
   Drupal.behaviors.ginEditForm = {
-    attach: (context) => {
+    attach: (context, settings) => {
       once('ginEditForm', '.region-content form.gin-node-edit-form', context).forEach(form => {
         const sticky = context.querySelector('.gin-sticky');
         const newParent = context.querySelector('.region-sticky__items__inner');
@@ -16,10 +16,19 @@
           const actionButtons = newParent.querySelectorAll('button, input, select, textarea');
 
           if (actionButtons.length > 0) {
+            let element_id = el.getAttribute('id');
+            let new_id = element_id + '--gin-edit-form';
+
             actionButtons.forEach((el) => {
               el.setAttribute('form', form.getAttribute('id'));
-              el.setAttribute('id', el.getAttribute('id') + '--gin-edit-form');
+              el.setAttribute('id', new_id);
             });
+
+            if (settings.ajax && settings.ajax.hasOwnProperty(element_id)) {
+              settings.ajax[new_id] = settings.ajax[element_id];
+              settings.ajax[new_id].selector = '#' + new_id;
+              delete settings.ajax[element_id];
+            }
           }
         }
       });
